<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Importing alevin scRNA-seq counts into R/Bioconductor • alevin2bioc</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Importing alevin scRNA-seq counts into R/Bioconductor">
<meta property="og:description" content="alevin2bioc">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">alevin2bioc</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.23</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/alevin2bioc.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/mikelove/alevin2bioc">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="alevin2bioc_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Importing alevin scRNA-seq counts into R/Bioconductor</h1>
                        <h4 class="author">Michael I. Love, UNC-Chapel Hill, <a href="https://mikelove.github.io">website</a>
</h4>
                        <h4 class="author">Avi Srivastava, New York Genome Center, <a href="https://k3yavi.github.io">website</a>
</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/mikelove/alevin2bioc/blob/master/vignettes/alevin2bioc.Rmd"><code>vignettes/alevin2bioc.Rmd</code></a></small>
      <div class="hidden name"><code>alevin2bioc.Rmd</code></div>

    </div>

    
    
<div id="workshop-description" class="section level2">
<h2 class="hasAnchor">
<a href="#workshop-description" class="anchor"></a>Workshop Description</h2>
<p>In this workshop, we will demonstrate basics of quantification of droplet-based scRNA-seq reads using <em>alevin</em>, producing a count matrix for import into Bioconductor using <em>tximeta</em>, in the end producing a <em>SingleCellExperiment</em> object. We will also demonstrate the ability of alevin to provide quantification uncertainty on the count matrix, and visualize this uncertainty across cells.</p>
<p>We plan the workshop to be an instructor-led live demo with time for questions and interactions with the participants. We imagine that the target participant for the workshop probably has some dscRNA-seq data, and knows about e.g. generating a count matrix with <em>CellRanger</em>. We will show an alternative quantification pipeline and explain its benefits. We will show how to hand off the data object to <a href="https://osca.bioconductor.org/">common single-cell workflows in Bioconductor (OSCA)</a> as well as to <a href="https://satijalab.org/seurat/">Seurat</a>.</p>
<div id="pre-requisites" class="section level3">
<h3 class="hasAnchor">
<a href="#pre-requisites" class="anchor"></a>Pre-requisites</h3>
<ul>
<li>Basic knowledge of R syntax</li>
<li>General understanding of scRNA-seq experiment</li>
</ul>
</div>
<div id="workshop-participation" class="section level3">
<h3 class="hasAnchor">
<a href="#workshop-participation" class="anchor"></a>Workshop Participation</h3>
<p>Students will participate by following along a live demo, and asking questions or providing feedback throughout.</p>
</div>
<div id="r-bioconductor-packages-used" class="section level3">
<h3 class="hasAnchor">
<a href="#r-bioconductor-packages-used" class="anchor"></a><em>R</em> / <em>Bioconductor</em> packages used</h3>
<ul>
<li>tximeta</li>
<li>SingleCellExperiment</li>
<li>fishpond</li>
<li>scran</li>
<li>Seurat</li>
</ul>
</div>
<div id="time-outline" class="section level3">
<h3 class="hasAnchor">
<a href="#time-outline" class="anchor"></a>Time outline</h3>
<p>An example for a 45-minute workshop:</p>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>alevin for droplet scRNA-seq</td>
<td>20m</td>
</tr>
<tr class="even">
<td>importing counts into Bioc</td>
<td>5m</td>
</tr>
<tr class="odd">
<td>examination of object &amp; counts</td>
<td>10m</td>
</tr>
<tr class="even">
<td>examination of uncertainty</td>
<td>10m</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="workshop-goals-and-objectives" class="section level2">
<h2 class="hasAnchor">
<a href="#workshop-goals-and-objectives" class="anchor"></a>Workshop goals and objectives</h2>
<div id="learning-goals" class="section level3">
<h3 class="hasAnchor">
<a href="#learning-goals" class="anchor"></a>Learning goals</h3>
<ul>
<li>understand how scRNA-seq quantification methods work and understanding their limits</li>
<li>describe how Bioconductor’s classes including <em>SingleCellExperiment</em> facilitate reproducibility through tracking metadata on the samples/cells and the genomic ranges</li>
</ul>
</div>
<div id="learning-objectives" class="section level3">
<h3 class="hasAnchor">
<a href="#learning-objectives" class="anchor"></a>Learning objectives</h3>
<ul>
<li>see code to run <em>alevin</em>, quantifying scRNA-seq reads to make a gene count matrix</li>
<li>import scRNA-seq count data including genomic ranges</li>
<li>manipulate a SingleCellExperiment</li>
<li>examine scRNA-seq counts over cell labels</li>
<li>examine uncertainty estimates for counts</li>
<li>hand-off to OSCA workflows</li>
<li>hand-off to Seurat workflow</li>
</ul>
</div>
</div>
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p><em>alevin</em> is a fast end-to-end pipeline to process droplet-based single-cell RNA sequencing (dscRNA-seq) data, generating per-cell gene count estimates by performing cell barcode detection, read mapping, and unique molecular identifier (UMI) deduplication.<span class="citation">(Srivastava et al. 2019)</span>. It extends the methods in the <em>Salmon</em> software <span class="citation">(Patro et al. 2017)</span>, and is distributed as part of <em>Salmon</em>.</p>
<p>The data we will use today is generated from the peripheral blood mononuclear cells (PBMCs) of a healthy donor which have been sequenced and made public by 10x Genomics (referred as PBMC_1k V2 Chemistry) <span class="citation">(“1k Pbmcs from a Healthy Donor (V2 Chemistry),” n.d.)</span>. These cells are well characterized, and so we will examine counts for cells grouped into cell types labelled by marker genes.</p>
</div>
<div id="running-alevin" class="section level2">
<h2 class="hasAnchor">
<a href="#running-alevin" class="anchor"></a>Running alevin</h2>
<p>In order to run <em>alevin</em>, we must first generate an index of the reference sequences. For this experiment, we used the latest <a href="https://www.gencodegenes.org/">GENCODE</a> human reference transcripts <span class="citation">(Frankish, GENCODE-consoritum, and Flicek 2018)</span>, which happen to be version 33 (in April 2020 version 34 was released). The index of the reference transcripts (FASTA) and a file linking transcripts to genes can be created following the instructions <a href="https://combine-lab.github.io/alevin-tutorial/2018/setting-up-resources/">here</a>. (An alternative is to also include genome sequence that does not fall into transcripts as part of the index. Details for creating such an index are provided <a href="https://combine-lab.github.io/alevin-tutorial/2019/selective-alignment/">here</a>.)</p>
<p>Finally, running <em>alevin</em> once an index has been created is just a single command, and detailed instructions can be found <a href="https://combine-lab.github.io/alevin-tutorial/2018/running-alevin/">here</a>. In this case, <code>--chromiumV3</code> was used and <code>--numCellBootstraps 30</code> in order to generate bootstrap inferential replicates. A simplified version of the command is shown below:</p>
<pre><code>salmon alevin -l ISR \
  -1 sample_L001_R1_001.fastq.gz \
  -2 sample_L001_R2_001.fastq.gz \
  --chromiumV3 -i index -p 12 -o sample --tgMap txp2gene.tsv \
  --numCellBootstraps 30</code></pre>
<p>Here, <code>-1</code> gives the CB+UMI file and <code>-2</code> gives the read sequence file. Multiple files can be provided to the arguments <code>-1</code> and <code>-2</code> with a single space between the files, as long as they are given in the same order to the two arguments.</p>
<p>Even more details about <em>alevin</em> arguments can be found <a href="https://salmon.readthedocs.io/en/latest/alevin.html">here</a>.</p>
<p><img width="600" src="alevin.gif"></p>
<p>For the rest of the tutorial we will focus on importing data into R/Bioconductor. The output directory of running <em>alevin</em> on the PBMC sequence data is included in this workflow package in the <code>extdata</code> directory. <em>alevin</em> took about 14 minutes to quantify 33 million reads from around 1700 cells. 30 bootstrap inferential replicates were generated, summarized to sparse inferential mean and variance matrices (discussed later).</p>
</div>
<div id="importing-alevin-data-with-tximeta" class="section level2">
<h2 class="hasAnchor">
<a href="#importing-alevin-data-with-tximeta" class="anchor"></a>Importing alevin data with tximeta</h2>
<p>We will use <em>tximeta</em> to import the <em>alevin</em> counts into R/Bioconductor. The main function <code>tximeta</code> reads information from the entire output directory of <em>alevin</em> or <em>Salmon</em> in order to automatically detect and download metadata about the reference sequences (the transcripts) <span class="citation">(Love et al. 2020)</span>. It should work “out of the box” for human, mouse, and fruit fly reference transcripts from GENCODE, Ensembl, or RefSeq</p>
<p>First we specify the path where the quantification data is stored. In this tutorial, the data is stored in an R package, so we need to use the <code>system.file</code> command.</p>
<p><strong>Note:</strong> For typical use, you would not use <code>system.file</code>, but would just specify the path to the directory of the output from <em>alevin</em>.</p>
<p><strong>Second note:</strong> when sharing <em>alevin</em> or <em>Salmon</em> datasets, make sure to share the entire output directory (you can <code>zip</code> or <code>tar</code> the directories to make them into a single share-able file).</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="co"># normally you just would use:</span>
<span class="co"># dir &lt;- "/path/to/alevin/output"</span>
<span class="no">extdata</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="kw">package</span><span class="kw">=</span><span class="st">"alevin2bioc"</span>)
<span class="no">dir</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">extdata</span>, <span class="st">"pbmc_1k"</span>)</pre></body></html></div>
<p>We make a vector called <code>files</code> (for <em>alevin</em> this will point to just a single file, but with <em>Salmon</em> usually one would import multiple files at a time).</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="no">files</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">dir</span>, <span class="st">"alevin"</span>, <span class="st">"quants_mat.gz"</span>)
<span class="fu"><a href="https://rdrr.io/r/base/files.html">file.exists</a></span>(<span class="no">files</span>)</pre></body></html></div>
<pre><code>## [1] TRUE</code></pre>
<p>We can import the <em>alevin</em> quantification using the following call to <code>tximeta</code>. The extra argument to <em>alevin</em> will filter out cells based on <em>alevin</em>’s post-quantification quality control methods (see paper for details).</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">tximeta</span>)
<span class="no">se</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/tximeta/man/tximeta.html">tximeta</a></span>(<span class="no">files</span>, <span class="kw">type</span><span class="kw">=</span><span class="st">"alevin"</span>, <span class="kw">alevinArgs</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">filterBarcodes</span><span class="kw">=</span><span class="fl">TRUE</span>))</pre></body></html></div>
<pre><code>## importing quantifications</code></pre>
<pre><code>## reading in alevin gene-level counts across cells with fishpond</code></pre>
<pre><code>## filtering down to 1142 cell barcodes</code></pre>
<pre><code>## reading in alevin inferential variance with fishpond</code></pre>
<pre><code>## found matching transcriptome:
## [ GENCODE - Homo sapiens - release 33 ]</code></pre>
<pre><code>## building TxDb with 'GenomicFeatures' package</code></pre>
<pre><code>## Import genomic features from the file as a GRanges object ... OK
## Prepare the 'metadata' data frame ... OK
## Make the TxDb object ...</code></pre>
<pre><code>## Warning in .get_cds_IDX(mcols0$type, mcols0$phase): The "phase" metadata column contains non-NA values for features of type
##   stop_codon. This information was ignored.</code></pre>
<pre><code>## OK
## generating gene ranges
## generating gene ranges
## fetching genome info for GENCODE</code></pre>
<p>We can see as it was importing that it automatically detected we are working with data quantified using GENCODE’s <em>Homo sapiens</em> reference transcripts, release 33. The transcript metadata was added programmatically to the output (this object, <code>se</code>). <code>tximeta</code> is also written in such a way that it will avoid unnecessary downloads and parsing of files – it will first attempt to find a cached version of the metadata, to avoid re-downloading or re-parsing these files.</p>
<p><code>tximeta</code> returns a <em>SummarizedExperiment</em> <span class="citation">(Lawrence et al. 2013)</span>. We can easily convert this object into a <em>SingleCellExperiment</em> <span class="citation">(Amezquita et al. 2020)</span> which has specific slots designed for single-cell experiment data.</p>
<div class="sourceCode" id="cb15"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/message.html">suppressPackageStartupMessages</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">SingleCellExperiment</span>))
<span class="no">sce</span> <span class="kw">&lt;-</span> <span class="fu">as</span>(<span class="no">se</span>, <span class="st">"SingleCellExperiment"</span>)</pre></body></html></div>
<p>The <em>SingleCellExperiment</em> object is used widely across Bioconductor packages (as you may already know), and so the code up to this point can be used as an entry point into other Bioconductor single-cell workflows. For more details on working with <em>SingleCellExperiment</em> objects, one can consult the following online book: <a href="https://osca.bioconductor.org/">Orchestrating Single-Cell Analysis with Bioconductor</a> <span class="citation">(Amezquita et al. 2020)</span>.</p>
<p>The data is now available as assays in <code>sce</code>. We can see what is available:</p>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="fu">assayNames</span>(<span class="no">sce</span>)</pre></body></html></div>
<pre><code>## [1] "counts"   "variance" "mean"</code></pre>
<p>And we can access individual gene-by-cell data by pulling out a particular assay. Note that, due to the use of EM (which avoids discarding multi-mapping reads), we will have some fractional counts in the counts matrix.</p>
<div class="sourceCode" id="cb18"><html><body><pre class="r"><span class="fu">assays</span>(<span class="no">sce</span>)<span class="kw">[[</span><span class="st">"counts"</span>]][<span class="fl">1</span>:<span class="fl">20</span>,<span class="fl">1</span>:<span class="fl">4</span>]</pre></body></html></div>
<pre><code>## 20 x 4 sparse Matrix of class "dgCMatrix"
##                   AAGACAACAGATCACT CAGGTATGTCAGTCTA TATCTTGTCCATCTCG
## ENSG00000223972.5                .                .                .
## ENSG00000227232.5                .                .                .
## ENSG00000278267.1                .                .                .
## ENSG00000243485.5                .                .                .
## ENSG00000284332.1                .                .                .
## ENSG00000237613.2                .                .                .
## ENSG00000268020.3                .                .                .
## ENSG00000240361.2                .                .                .
## ENSG00000186092.6                .                .                .
## ENSG00000238009.6                .                .                .
## ENSG00000239945.1                .                .                .
## ENSG00000233750.3                .                .                .
## ENSG00000268903.1                .                .                .
## ENSG00000269981.1                .                .                .
## ENSG00000239906.1                .                .                .
## ENSG00000241860.7                1                .                .
## ENSG00000222623.1                .                .                .
## ENSG00000241599.1                .                .                .
## ENSG00000279928.2                .                .                .
## ENSG00000279457.4                .                .                .
##                   TTGTTCACACTTGTGA
## ENSG00000223972.5                .
## ENSG00000227232.5                .
## ENSG00000278267.1                .
## ENSG00000243485.5                .
## ENSG00000284332.1                .
## ENSG00000237613.2                .
## ENSG00000268020.3                .
## ENSG00000240361.2                .
## ENSG00000186092.6                .
## ENSG00000238009.6                .
## ENSG00000239945.1                .
## ENSG00000233750.3                .
## ENSG00000268903.1                .
## ENSG00000269981.1                .
## ENSG00000239906.1                .
## ENSG00000241860.7                .
## ENSG00000222623.1                .
## ENSG00000241599.1                .
## ENSG00000279928.2                .
## ENSG00000279457.4                .</code></pre>
<p>For the <code>counts</code> matrix, we can also use the <code><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html">counts()</a></code> accessor function:</p>
<div class="sourceCode" id="cb20"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html">counts</a></span>(<span class="no">sce</span>)[,<span class="fl">1</span>:<span class="fl">4</span>])</pre></body></html></div>
<pre><code>## AAGACAACAGATCACT CAGGTATGTCAGTCTA TATCTTGTCCATCTCG TTGTTCACACTTGTGA 
##            19030            24657            29921            25864</code></pre>
</div>
<div id="benefits-of-tximeta" class="section level2">
<h2 class="hasAnchor">
<a href="#benefits-of-tximeta" class="anchor"></a>Benefits of tximeta</h2>
<p>We can automatically add gene IDs, because <em>tximeta</em> knows the type of identifiers on the rows of the <code>sce</code> object:</p>
<div class="sourceCode" id="cb22"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">org.Hs.eg.db</span>) <span class="co"># org pkg for Homo sapiens</span></pre></body></html></div>
<pre><code>## </code></pre>
<div class="sourceCode" id="cb24"><html><body><pre class="r"><span class="no">sce</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/tximeta/man/addIds.html">addIds</a></span>(<span class="no">sce</span>, <span class="st">"SYMBOL"</span>)</pre></body></html></div>
<pre><code>## mapping to new IDs using 'org.Hs.eg.db' data package
## if all matching IDs are desired, and '1:many mappings' are reported,
## set multiVals='list' to obtain all the matching IDs</code></pre>
<pre><code>## it appears the rows are gene IDs, setting 'gene' to TRUE</code></pre>
<pre><code>## 'select()' returned 1:many mapping between keys and columns</code></pre>
<div class="sourceCode" id="cb28"><html><body><pre class="r"><span class="fu">mcols</span>(<span class="no">sce</span>)</pre></body></html></div>
<pre><code>## DataFrame with 60233 rows and 2 columns
##                             gene_id      SYMBOL
##                         &lt;character&gt; &lt;character&gt;
## ENSG00000223972.5 ENSG00000223972.5     DDX11L1
## ENSG00000227232.5 ENSG00000227232.5          NA
## ENSG00000278267.1 ENSG00000278267.1   MIR6859-1
## ENSG00000243485.5 ENSG00000243485.5          NA
## ENSG00000284332.1 ENSG00000284332.1   MIR1302-2
## ...                             ...         ...
## ENSG00000198695.2 ENSG00000198695.2         ND6
## ENSG00000210194.1 ENSG00000210194.1          NA
## ENSG00000198727.2 ENSG00000198727.2        CYTB
## ENSG00000210195.2 ENSG00000210195.2          NA
## ENSG00000210196.2 ENSG00000210196.2          NA</code></pre>
<p>Also, because the provenance was detected, we also have the ranges of the genes in their proper genomic context. So it is easy to find, for example, genes near a particular position in the genome, in this case 4 genes that overlap the range <code>chr1:10,000,000-10,100,000</code>.</p>
<div class="sourceCode" id="cb30"><html><body><pre class="r"><span class="no">x</span> <span class="kw">&lt;-</span> <span class="fu">GRanges</span>(<span class="st">"chr1"</span>, <span class="fu">IRanges</span>(<span class="fl">10e6</span>,<span class="fl">10.1e6</span>))
<span class="no">sce</span>[<span class="no">sce</span> <span class="kw">%over%</span> <span class="no">x</span>,]</pre></body></html></div>
<pre><code>## class: SingleCellExperiment 
## dim: 4 1142 
## metadata(6): tximetaInfo quantInfo ... txomeInfo txdbInfo
## assays(3): counts variance mean
## rownames(4): ENSG00000162444.12 ENSG00000130939.20 ENSG00000224340.1
##   ENSG00000233623.2
## rowData names(2): gene_id SYMBOL
## colnames(1142): AAGACAACAGATCACT CAGGTATGTCAGTCTA ... CTCTGGTAGTAGGATT
##   AGGAGGTAGGTCACAG
## colData names(0):
## reducedDimNames(0):
## altExpNames(0):</code></pre>
</div>
<div id="add-cell-annotations" class="section level2">
<h2 class="hasAnchor">
<a href="#add-cell-annotations" class="anchor"></a>Add cell annotations</h2>
<p>Cell annotations were already generated using <em>Seurat</em> <span class="citation">(Stuart et al. 2019)</span>. The script is saved in this package in <code>inst/scripts/seurat.R</code>. Here we will use them for size factor estimation and for visualization.</p>
<div class="sourceCode" id="cb32"><html><body><pre class="r"><span class="no">ids</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">extdata</span>, <span class="st">"idents.rds"</span>))
<span class="no">top10</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">extdata</span>, <span class="st">"top10.csv"</span>))</pre></body></html></div>
<p>We subset to the cells that we have IDs for, and attach the labels in the correct order:</p>
<div class="sourceCode" id="cb33"><html><body><pre class="r"><span class="no">idx</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">sce</span>) <span class="kw">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">ids</span>)
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">idx</span>)</pre></body></html></div>
<pre><code>## idx
## FALSE  TRUE 
##    74  1068</code></pre>
<div class="sourceCode" id="cb35"><html><body><pre class="r"><span class="no">sce</span> <span class="kw">&lt;-</span> <span class="no">sce</span>[,<span class="no">idx</span>]
<span class="no">sce</span>$<span class="no">cluster</span> <span class="kw">&lt;-</span> <span class="no">ids</span>[<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">sce</span>)]</pre></body></html></div>
<p>The number of cells per cluster:</p>
<div class="sourceCode" id="cb36"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">sce</span>$<span class="no">cluster</span>)</pre></body></html></div>
<pre><code>## 
##           CD4 T CD14+ Monocytes               B           CD8 T              NK 
##             346             346             189             133              54</code></pre>
<p>Note that the different clusters have different total counts, for example:</p>
<div class="sourceCode" id="cb38"><html><body><pre class="r"><span class="no">cs</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span>(<span class="fu">assays</span>(<span class="no">sce</span>)<span class="kw">[[</span><span class="st">"counts"</span>]])
<span class="co"># cells with higher number of UMI</span>
<span class="no">more.umi</span> <span class="kw">&lt;-</span> <span class="no">cs</span> <span class="kw">&gt;</span> <span class="fl">10000</span>
(<span class="no">tab</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">more.umi</span>, <span class="no">sce</span>$<span class="no">cluster</span>))</pre></body></html></div>
<pre><code>##         
## more.umi CD4 T CD14+ Monocytes   B CD8 T  NK
##    FALSE   322             159 165   130  54
##    TRUE     24             187  24     3   0</code></pre>
<div class="sourceCode" id="cb40"><html><body><pre class="r"><span class="fl">100</span> * <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/proportions.html">prop.table</a></span>(<span class="no">tab</span>,<span class="fl">2</span>),<span class="fl">2</span>) <span class="co"># percent</span></pre></body></html></div>
<pre><code>##         
## more.umi CD4 T CD14+ Monocytes   B CD8 T  NK
##    FALSE    93              46  87    98 100
##    TRUE      7              54  13     2   0</code></pre>
<div class="sourceCode" id="cb42"><html><body><pre class="r"><span class="co"># cell with lower number of UMI</span>
<span class="no">fewer.umi</span> <span class="kw">&lt;-</span> <span class="no">cs</span> <span class="kw">&lt;</span> <span class="fl">5000</span>
(<span class="no">tab</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">fewer.umi</span>, <span class="no">sce</span>$<span class="no">cluster</span>))</pre></body></html></div>
<pre><code>##          
## fewer.umi CD4 T CD14+ Monocytes   B CD8 T  NK
##     FALSE   316             300 149   107  39
##     TRUE     30              46  40    26  15</code></pre>
<div class="sourceCode" id="cb44"><html><body><pre class="r"><span class="fl">100</span> * <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/proportions.html">prop.table</a></span>(<span class="no">tab</span>,<span class="fl">2</span>),<span class="fl">2</span>) <span class="co"># percent</span></pre></body></html></div>
<pre><code>##          
## fewer.umi CD4 T CD14+ Monocytes  B CD8 T NK
##     FALSE    91              87 79    80 72
##     TRUE      9              13 21    20 28</code></pre>
<p>We have a data.frame with the top marker genes per cluster, as identified with <em>Seurat</em>.</p>
<div class="sourceCode" id="cb46"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">top10</span>)</pre></body></html></div>
<pre><code>##   X         p_val avg_logFC pct.1 pct.2     p_val_adj cluster
## 1 1 3.399988e-135 1.2134065 0.997 0.602 6.075778e-131       0
## 2 2 2.755855e-116 1.0112296 0.879 0.177 4.924713e-112       0
## 3 3 4.214635e-114 0.9936307 0.968 0.194 7.531553e-110       0
## 4 4 1.644166e-111 0.9595734 0.855 0.154 2.938125e-107       0
## 5 5 9.344999e-110 1.0665597 0.980 0.266 1.669951e-105       0
## 6 6 2.602057e-105 1.1441765 0.916 0.175 4.649875e-101       0
##                 gene symbol cluster.name
## 1 ENSG00000111716.14   LDHB        CD4 T
## 2 ENSG00000081059.20   TCF7        CD4 T
## 3  ENSG00000167286.9   CD3D        CD4 T
## 4 ENSG00000127152.18 BCL11B        CD4 T
## 5  ENSG00000277734.8   &lt;NA&gt;        CD4 T
## 6 ENSG00000168685.15   IL7R        CD4 T</code></pre>
</div>
<div id="plotting-counts-with-uncertainty" class="section level2">
<h2 class="hasAnchor">
<a href="#plotting-counts-with-uncertainty" class="anchor"></a>Plotting counts with uncertainty</h2>
<p>In this section, we will begin to plot the counts for cells, for specific genes, and showing the <em>inferential uncertainty</em> as quantified by <em>alevin</em>. This is a unique aspect to the <em>alevin</em> quantification method, that it uses EM to assign gene multi-mapping reads, instead of discarding these reads. Note that many reads can be discarded with alternative pipelines, and these are not uniformly lost from all genes, but the number of multi-mapping reads is higher for gene families (i.e. genes with high sequence homology). See the <em>alevin</em> publication for more details on this aspect of bias in dscRNA-seq counts <span class="citation">(Srivastava et al. 2019)</span>.</p>
<p>To get a systematic sense of how many reads are preserved by using EM-based assignment of multi-mapping, consider the following plot constructed using a recent dataset of mouse embryos <span class="citation">(Pijuan-Sala et al. 2019)</span>. Focusing on genes where the count changes by more than 50%, the left side shows the total count per gene (summed over cells) without EM, and the right side shows the gene counts using <em>alevin</em> with EM.</p>
<p><img width="400" src="em_plot.png"></p>
<p><em>alevin</em> can also attach a measure of uncertainty to each count in the matrix. <em>alevin</em> computes the mean and variance of <em>inferential replicates</em> which are generated by bootstrapping the read data. <code>tximeta</code> will import these inferential mean and variance matrices by default (they are also sparse, similar to the counts matrix). The developers of <em>alevin</em> and <em>fishpond</em> have recently examined the use of the compressed inferential uncertainty matrices in various single-cell analysis tasks, including DE and trajectory analysis <span class="citation">(Van Buren et al. 2020)</span>.</p>
<p>We will first visualize the uncertainty, and later give an example of a set of genes where the uncertainty is indicating significant quantification uncertainty which is useful to consider when performing EDA (exploratory data analysis) or statistical analysis.</p>
<p>For a later demonstration of scaling, we will sort the cells by the total count (this is not something you would necessarily do in a typical analysis).</p>
<div class="sourceCode" id="cb48"><html><body><pre class="r"><span class="no">o</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span>(<span class="fu">assays</span>(<span class="no">sce</span>)<span class="kw">[[</span><span class="st">"counts"</span>]]), <span class="kw">decreasing</span><span class="kw">=</span><span class="fl">TRUE</span>)
<span class="no">sce</span> <span class="kw">&lt;-</span> <span class="no">sce</span>[,<span class="no">o</span>]</pre></body></html></div>
<p>We can now use a plotting function <code>plotInfReps</code> from the <em>fishpond</em> package in order to plot the inferential mean and variance for the cells, for various genes.</p>
<div class="sourceCode" id="cb49"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">fishpond</span>)
<span class="fu"><a href="https://rdrr.io/pkg/fishpond/man/plotInfReps.html">plotInfReps</a></span>(<span class="no">sce</span>, <span class="kw">idx</span><span class="kw">=</span><span class="st">"ENSG00000167286.9"</span>,
            <span class="kw">x</span><span class="kw">=</span><span class="st">"cluster"</span>, <span class="kw">mainCol</span><span class="kw">=</span><span class="st">"SYMBOL"</span>,
            <span class="kw">legend</span><span class="kw">=</span><span class="fl">TRUE</span>)</pre></body></html></div>
<p><img src="alevin2bioc_files/figure-html/plot-basic-1.png" width="700"></p>
<p>Here, we have <code>x</code> as a grouping variable for coloring the points, one can also specify <code>x</code> to be a continuous covariate, e.g. pseudo-time. An additional covariate <code>cov</code> can be used for additional grouping, e.g. batches, donors or lineages.</p>
<p>Here we show the same plot but now subsetting the number of cells:</p>
<div class="sourceCode" id="cb50"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">1</span>)
<span class="no">idx</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(<span class="no">sce</span>),<span class="fl">200</span>)
<span class="fu"><a href="https://rdrr.io/pkg/fishpond/man/plotInfReps.html">plotInfReps</a></span>(<span class="no">sce</span>[,<span class="no">idx</span>],
            <span class="kw">idx</span><span class="kw">=</span><span class="st">"ENSG00000167286.9"</span>,
            <span class="kw">x</span><span class="kw">=</span><span class="st">"cluster"</span>, <span class="kw">mainCol</span><span class="kw">=</span><span class="st">"SYMBOL"</span>,
            <span class="kw">legend</span><span class="kw">=</span><span class="fl">TRUE</span>)</pre></body></html></div>
<p><img src="alevin2bioc_files/figure-html/plot-medium-1.png" width="700"></p>
<p>This time only 100 cells:</p>
<div class="sourceCode" id="cb51"><html><body><pre class="r"><span class="no">idx</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(<span class="no">sce</span>),<span class="fl">100</span>)
<span class="fu"><a href="https://rdrr.io/pkg/fishpond/man/plotInfReps.html">plotInfReps</a></span>(<span class="no">sce</span>[,<span class="no">idx</span>],
            <span class="kw">idx</span><span class="kw">=</span><span class="st">"ENSG00000167286.9"</span>,
            <span class="kw">x</span><span class="kw">=</span><span class="st">"cluster"</span>, <span class="kw">mainCol</span><span class="kw">=</span><span class="st">"SYMBOL"</span>,
            <span class="kw">legend</span><span class="kw">=</span><span class="fl">TRUE</span>)</pre></body></html></div>
<p><img src="alevin2bioc_files/figure-html/plot-small-1.png" width="700"></p>
<p>The plots change their features across different sample sizes for ease of viewing the uncertainty of quantification for individual cells.</p>
<p>We can also choose to plot the cells in their original order (the default for scRNA-seq is to sort by the mean value):</p>
<div class="sourceCode" id="cb52"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/fishpond/man/plotInfReps.html">plotInfReps</a></span>(<span class="no">sce</span>, <span class="kw">idx</span><span class="kw">=</span><span class="st">"ENSG00000167286.9"</span>,
            <span class="kw">x</span><span class="kw">=</span><span class="st">"cluster"</span>, <span class="kw">mainCol</span><span class="kw">=</span><span class="st">"SYMBOL"</span>,
            <span class="kw">reorder</span><span class="kw">=</span><span class="fl">FALSE</span>)</pre></body></html></div>
<p><img src="alevin2bioc_files/figure-html/plot-no-order-1.png" width="700"></p>
</div>
<div id="scaling-with-size-factors" class="section level2">
<h2 class="hasAnchor">
<a href="#scaling-with-size-factors" class="anchor"></a>Scaling with size factors</h2>
<p>We use <code>computeSumFactors</code> <span class="citation">(Lun, Bach, and Marioni 2016)</span> from the <em>scran</em> package <span class="citation">(A. T. L. Lun, McCarthy, and Marioni 2016)</span> to compute size factors that are stored in <code><a href="https://rdrr.io/pkg/SingleCellExperiment/man/sizeFactors.html">sizeFactors(sce)</a></code>.</p>
<div class="sourceCode" id="cb53"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">scran</span>)
<span class="no">sce</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/computeSumFactors.html">computeSumFactors</a></span>(<span class="no">sce</span>, <span class="kw">clusters</span><span class="kw">=</span><span class="no">sce</span>$<span class="no">cluster</span>)
<span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="no">cs</span>, <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/sizeFactors.html">sizeFactors</a></span>(<span class="no">sce</span>), <span class="kw">xlab</span><span class="kw">=</span><span class="st">"column sum"</span>, <span class="kw">ylab</span><span class="kw">=</span><span class="st">"sum factor"</span>)</pre></body></html></div>
<p><img src="alevin2bioc_files/figure-html/size-factors-1.png" width="700"></p>
<p>Now we demonstrate un-scaled counts and counts scaling with size factors just computed. (Note in the second plot that the first cell in each group isn’t the highest anymore.)</p>
<div class="sourceCode" id="cb54"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span>(<span class="kw">mfrow</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">2</span>,<span class="fl">1</span>), <span class="kw">mar</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">2.5</span>,<span class="fl">4.5</span>,<span class="fl">1</span>,<span class="fl">1</span>))
<span class="fu"><a href="https://rdrr.io/pkg/fishpond/man/plotInfReps.html">plotInfReps</a></span>(<span class="no">sce</span>, <span class="kw">idx</span><span class="kw">=</span><span class="st">"ENSG00000167286.9"</span>,
            <span class="kw">x</span><span class="kw">=</span><span class="st">"cluster"</span>, <span class="kw">main</span><span class="kw">=</span><span class="st">""</span>,
            <span class="kw">reorder</span><span class="kw">=</span><span class="fl">FALSE</span>)
<span class="fu"><a href="https://rdrr.io/pkg/fishpond/man/plotInfReps.html">plotInfReps</a></span>(<span class="no">sce</span>, <span class="kw">idx</span><span class="kw">=</span><span class="st">"ENSG00000167286.9"</span>,
            <span class="kw">x</span><span class="kw">=</span><span class="st">"cluster"</span>, <span class="kw">main</span><span class="kw">=</span><span class="st">""</span>,
            <span class="kw">applySF</span><span class="kw">=</span><span class="fl">TRUE</span>, <span class="kw">reorder</span><span class="kw">=</span><span class="fl">FALSE</span>)</pre></body></html></div>
<p><img src="alevin2bioc_files/figure-html/scaling-1.png" width="700"></p>
</div>
<div id="inferential-variance-uncertainty" class="section level2">
<h2 class="hasAnchor">
<a href="#inferential-variance-uncertainty" class="anchor"></a>Inferential variance (uncertainty)</h2>
<p>In this second to last section, we will briefly talk about how the inferential uncertainty as stored in the variance assay may be useful in practice.</p>
<p>Many users may prefer to just work with the counts matrix, and not consider the inferential mean and variance. We have found that, globally, this may not lead to too much of a problem, but for certain genes, it may be important to use the inferential variance in cases where it may signal difficult to quantify genes. Inferential uncertainty for bulk and single-cell RNA-seq differential expression was a focus of the <em>Swish</em> nonparametric statistical method <span class="citation">(Zhu et al. 2019)</span> which we do not demonstrate here, but one can refer to the vignette of <a href="https://bioconductor.org/packages/swish">swish</a> for more details (in particular the section on <em>alevin</em> data).</p>
<p>Let’s start by visualizing the uncertainty across all values in the matrix:</p>
<div class="sourceCode" id="cb55"><html><body><pre class="r"><span class="no">var</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html">as.vector</a></span>(<span class="fu">assays</span>(<span class="no">sce</span>)<span class="kw">[[</span><span class="st">"variance"</span>]])
<span class="no">mu</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html">as.vector</a></span>(<span class="fu">assays</span>(<span class="no">sce</span>)<span class="kw">[[</span><span class="st">"mean"</span>]])
<span class="no">idx</span> <span class="kw">&lt;-</span> <span class="no">mu</span> <span class="kw">&gt;</span> <span class="fl">3</span>
<span class="no">df</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw">log10mean</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span>(<span class="no">mu</span>[<span class="no">idx</span>]),
                 <span class="kw">log10var</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span>(<span class="no">var</span>[<span class="no">idx</span>]))</pre></body></html></div>
<div class="sourceCode" id="cb56"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">ggplot2</span>)
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="no">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="no">log10mean</span>, <span class="no">log10var</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_hex.html">geom_hex</a></span>(<span class="kw">bins</span><span class="kw">=</span><span class="fl">100</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_abline</a></span>(<span class="kw">intercept</span><span class="kw">=</span><span class="fl">0</span>, <span class="kw">slope</span><span class="kw">=</span><span class="fl">1</span>, <span class="kw">col</span><span class="kw">=</span><span class="st">"red"</span>)</pre></body></html></div>
<p><img src="alevin2bioc_files/figure-html/var-mean-1.png" width="700"></p>
<p>The values around the red line indicate not much uncertainty (bootstrapping reads gives us roughly multinomial, and so also approximately Poisson, variability around the mean count). However we can see a tail of higher uncertainty values in the matrix, where the inferential variance is for example, up to 10 times higher than the mean.</p>
<p>We can also plot this summarized to a single value per gene, here we calculate the 99% quantile of the ratio of variance over mean, per gene, and plot this over the mean:</p>
<div class="sourceCode" id="cb57"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">matrixStats</span>)
<span class="no">rratio</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/matrixStats/man/rowQuantiles.html">rowQuantiles</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>(
  <span class="fu">assays</span>(<span class="no">sce</span>)<span class="kw">[[</span><span class="st">"variance"</span>]] /
  (<span class="fu">assays</span>(<span class="no">sce</span>)<span class="kw">[[</span><span class="st">"mean"</span>]] + <span class="fl">1</span>)), <span class="kw">probs</span><span class="kw">=</span><span class="fl">.99</span>)
<span class="no">rmu</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowMeans</a></span>(<span class="fu">assays</span>(<span class="no">sce</span>)<span class="kw">[[</span><span class="st">"mean"</span>]])
<span class="no">idx</span> <span class="kw">&lt;-</span> <span class="no">rmu</span> <span class="kw">&gt;</span> <span class="fl">.1</span>
<span class="no">df</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw">log10mean</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span>(<span class="no">rmu</span>[<span class="no">idx</span>]),
                 <span class="kw">rratio</span><span class="kw">=</span><span class="no">rratio</span>[<span class="no">idx</span>],
                 <span class="kw">gene</span><span class="kw">=</span><span class="fu">mcols</span>(<span class="no">sce</span>)$<span class="no">SYMBOL</span>[<span class="no">idx</span>])</pre></body></html></div>
<p>A number of genes have a ratio above 2, but then two genes in this dataset stand out above the rest, and they happen to be a ribosomal gene, and another gene that is a read-through of the same locus:</p>
<div class="sourceCode" id="cb58"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/with.html">with</a></span>(<span class="no">df</span>, <span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="no">log10mean</span>, <span class="no">rratio</span>))
<span class="no">high.uncert</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span>(<span class="no">df</span>$<span class="no">rratio</span> <span class="kw">&gt;</span> <span class="fl">10</span>)
<span class="fu"><a href="https://rdrr.io/r/base/with.html">with</a></span>(<span class="no">df</span>[<span class="no">high.uncert</span>,],
     <span class="fu"><a href="https://rdrr.io/r/graphics/points.html">points</a></span>(<span class="no">log10mean</span>, <span class="no">rratio</span>, <span class="kw">pch</span><span class="kw">=</span><span class="fl">20</span>, <span class="kw">col</span><span class="kw">=</span><span class="st">"red"</span>))
<span class="fu"><a href="https://rdrr.io/r/base/with.html">with</a></span>(<span class="no">df</span>[<span class="no">high.uncert</span>,],
     <span class="fu"><a href="https://rdrr.io/r/graphics/text.html">text</a></span>(<span class="no">log10mean</span>, <span class="no">rratio</span>, <span class="no">gene</span>, <span class="kw">pos</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">2</span>,<span class="fl">4</span>)))</pre></body></html></div>
<p><img src="alevin2bioc_files/figure-html/high-var-mean-ratio-1.png" width="700"></p>
<div class="sourceCode" id="cb59"><html><body><pre class="r"><span class="fu">mcols</span>(<span class="no">sce</span>)$<span class="no">SYMBOL</span>[<span class="no">idx</span>][<span class="no">high.uncert</span>]</pre></body></html></div>
<pre><code>##  ENSG00000215472  ENSG00000265681 
## "RPL17-C18orf32"          "RPL17"</code></pre>
<p>Some of the counts for the ribosomal gene and its read-through:</p>
<div class="sourceCode" id="cb61"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span>(<span class="kw">mfrow</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">2</span>,<span class="fl">1</span>), <span class="kw">mar</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">2.5</span>,<span class="fl">4.5</span>,<span class="fl">1</span>,<span class="fl">1</span>))
<span class="fu"><a href="https://rdrr.io/pkg/fishpond/man/plotInfReps.html">plotInfReps</a></span>(<span class="no">sce</span>[,<span class="fl">1</span>:<span class="fl">100</span>], <span class="kw">idx</span><span class="kw">=</span><span class="st">"ENSG00000265681.7"</span>,
            <span class="kw">x</span><span class="kw">=</span><span class="st">"cluster"</span>, <span class="kw">mainCol</span><span class="kw">=</span><span class="st">"SYMBOL"</span>)
<span class="fu"><a href="https://rdrr.io/pkg/fishpond/man/plotInfReps.html">plotInfReps</a></span>(<span class="no">sce</span>[,<span class="fl">1</span>:<span class="fl">100</span>], <span class="kw">idx</span><span class="kw">=</span><span class="st">"ENSG00000215472.10"</span>,
            <span class="kw">x</span><span class="kw">=</span><span class="st">"cluster"</span>, <span class="kw">mainCol</span><span class="kw">=</span><span class="st">"SYMBOL"</span>)</pre></body></html></div>
<p><img src="alevin2bioc_files/figure-html/plot-high-uncert-1.png" width="700"></p>
<p>The extra uncertainty on the counts for this gene indicate that <em>alevin</em> was not certain if the reads should come from the gene or its read-through. While in this case, the ribosomal gene may not be over interest, there are other cases (e.g. developmental or immune genes with high sequence homology) where information of the uncertainty of quantification can be useful in interpreting the data. Such examples are provided in a recent preprint from the <em>alevin</em> and <em>fishpond</em> team <span class="citation">(Van Buren et al. 2020)</span>.</p>
</div>
<div id="downstream-analysis-with-seurat" class="section level2">
<h2 class="hasAnchor">
<a href="#downstream-analysis-with-seurat" class="anchor"></a>Downstream analysis with Seurat</h2>
<p>As we previously showed how to construct a <em>SingleCellExperiment</em> which can be used with other Bioconductor workflows, we also demonstrate how it is easy to convert the <code>sce</code> object into an object for use with the <em>Seurat</em> R package <span class="citation">(Stuart et al. 2019)</span> for single-cell analysis. As we noted, <em>Seurat</em> was already used to identify the cell types (with the script stored in <code>inst/scripts/seurat.R</code>).</p>
<p>We now load <em>Seurat</em> and create a <em>Seurat</em> object:</p>
<div class="sourceCode" id="cb62"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">Seurat</span>)</pre></body></html></div>
<pre><code>## 
## Attaching package: 'Seurat'</code></pre>
<pre><code>## The following object is masked from 'package:SummarizedExperiment':
## 
##     Assays</code></pre>
<div class="sourceCode" id="cb65"><html><body><pre class="r"><span class="no">cts</span> <span class="kw">&lt;-</span> <span class="fu">assays</span>(<span class="no">sce</span>)<span class="kw">[[</span><span class="st">"counts"</span>]]
<span class="no">pbmc</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/CreateSeuratObject.html">CreateSeuratObject</a></span>(<span class="no">cts</span>)</pre></body></html></div>
<p>We can easily create violin plots, for example:</p>
<div class="sourceCode" id="cb66"><html><body><pre class="r"><span class="no">mt.genes</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">sce</span>)[<span class="fu"><a href="https://rdrr.io/r/base/logical.html">as.logical</a></span>(<span class="fu">seqnames</span>(<span class="no">sce</span>) <span class="kw">==</span> <span class="st">"chrM"</span>)]
<span class="no">pbmc</span><span class="kw">[[</span><span class="st">"percent.mt"</span>]] <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/PercentageFeatureSet.html">PercentageFeatureSet</a></span>(<span class="no">pbmc</span>, <span class="kw">features</span><span class="kw">=</span><span class="no">mt.genes</span>)
<span class="no">feats</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"nFeature_RNA"</span>, <span class="st">"nCount_RNA"</span>, <span class="st">"percent.mt"</span>)
<span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/VlnPlot.html">VlnPlot</a></span>(<span class="no">pbmc</span>, <span class="kw">features</span><span class="kw">=</span><span class="no">feats</span>, <span class="kw">ncol</span><span class="kw">=</span><span class="fl">3</span>)</pre></body></html></div>
<p><img src="alevin2bioc_files/figure-html/seurat-violin-1.png" width="700"></p>
<p>From this point, one can use the <code>pbmc</code> object for use in <em>Seurat</em> workflows, for example, the <a href="https://satijalab.org/seurat/vignettes.html">vignettes</a> on the <em>Seurat</em> website.</p>
</div>
<div id="support" class="section level2">
<h2 class="hasAnchor">
<a href="#support" class="anchor"></a>Support</h2>
<p>For support on this tutorial, feel free to post to <a href="https://support.bioconductor.org" class="uri">https://support.bioconductor.org</a> and tag the post with the appropriate package (e.g. <code>tximeta</code> if the question is particular to the <code>tximeta</code> import aspect). If you are asking about a particular function, don’t forget to first read the man page (e.g. <code><a href="https://rdrr.io/pkg/tximeta/man/tximeta.html">?tximeta</a></code>), and also check the relevant package vignette for relevant details (e.g. <a href="https://bioconductor.org/packages/release/bioc/vignettes/tximeta/inst/doc/tximeta.html">tximeta vignette</a>).</p>
<p>For questions about <em>alevin</em>, first consult the online documentation at these links</p>
<p><a href="https://combine-lab.github.io/alevin-tutorial/" class="uri">https://combine-lab.github.io/alevin-tutorial/</a></p>
<p><a href="https://salmon.readthedocs.io/en/latest/alevin.html" class="uri">https://salmon.readthedocs.io/en/latest/alevin.html</a></p>
<p>You can also find links for seeking further support here:</p>
<p><a href="https://github.com/COMBINE-lab/salmon" class="uri">https://github.com/COMBINE-lab/salmon</a></p>
</div>
<div id="session-info" class="section level2">
<h2 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h2>
<div class="sourceCode" id="cb67"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()</pre></body></html></div>
<pre><code>## R version 4.0.0 (2020-04-24)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 18.04.4 LTS
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] Seurat_3.2.0                ggplot2_3.3.2              
##  [3] scran_1.17.9                fishpond_1.5.35            
##  [5] org.Hs.eg.db_3.11.4         SingleCellExperiment_1.11.6
##  [7] SummarizedExperiment_1.19.6 DelayedArray_0.15.7        
##  [9] matrixStats_0.56.0          Matrix_1.2-18              
## [11] tximeta_1.7.6               GenomicFeatures_1.41.2     
## [13] AnnotationDbi_1.51.1        Biobase_2.49.0             
## [15] GenomicRanges_1.41.5        GenomeInfoDb_1.25.8        
## [17] IRanges_2.23.10             S4Vectors_0.27.12          
## [19] BiocGenerics_0.35.4        
## 
## loaded via a namespace (and not attached):
##   [1] reticulate_1.16               tidyselect_1.1.0             
##   [3] RSQLite_2.2.0                 htmlwidgets_1.5.1            
##   [5] grid_4.0.0                    BiocParallel_1.23.2          
##   [7] Rtsne_0.15                    munsell_0.5.0                
##   [9] codetools_0.2-16              ica_1.0-2                    
##  [11] statmod_1.4.34                future_1.18.0                
##  [13] miniUI_0.1.1.1                withr_2.2.0                  
##  [15] colorspace_1.4-1              knitr_1.29                   
##  [17] ROCR_1.0-11                   tensor_1.5                   
##  [19] listenv_0.8.0                 labeling_0.3                 
##  [21] tximport_1.17.2               GenomeInfoDbData_1.2.3       
##  [23] polyclip_1.10-0               bit64_0.9-7.1                
##  [25] farver_2.0.3                  rprojroot_1.3-2              
##  [27] vctrs_0.3.2                   generics_0.0.2               
##  [29] xfun_0.15                     BiocFileCache_1.13.0         
##  [31] R6_2.4.1                      rsvd_1.0.3                   
##  [33] locfit_1.5-9.4                AnnotationFilter_1.13.0      
##  [35] bitops_1.0-6                  spatstat.utils_1.17-0        
##  [37] assertthat_0.2.1              promises_1.1.1               
##  [39] scales_1.1.1                  gtable_0.3.0                 
##  [41] globals_0.12.5                goftest_1.2-2                
##  [43] ensembldb_2.13.1              rlang_0.4.7                  
##  [45] splines_4.0.0                 rtracklayer_1.49.3           
##  [47] lazyeval_0.2.2                hexbin_1.28.1                
##  [49] BiocManager_1.30.10           yaml_2.2.1                   
##  [51] reshape2_1.4.4                abind_1.4-5                  
##  [53] backports_1.1.8               httpuv_1.5.4                 
##  [55] tools_4.0.0                   ellipsis_0.3.1               
##  [57] RColorBrewer_1.1-2            ggridges_0.5.2               
##  [59] Rcpp_1.0.5                    plyr_1.8.6                   
##  [61] progress_1.2.2                zlibbioc_1.35.0              
##  [63] purrr_0.3.4                   RCurl_1.98-1.2               
##  [65] prettyunits_1.1.1             rpart_4.1-15                 
##  [67] openssl_1.4.2                 deldir_0.1-28                
##  [69] pbapply_1.4-2                 cowplot_1.0.0                
##  [71] zoo_1.8-8                     ggrepel_0.8.2                
##  [73] cluster_2.1.0                 fs_1.4.2                     
##  [75] magrittr_1.5                  data.table_1.12.8            
##  [77] lmtest_0.9-37                 RANN_2.6.1                   
##  [79] ProtGenerics_1.21.0           fitdistrplus_1.1-1           
##  [81] hms_0.5.3                     patchwork_1.0.1              
##  [83] mime_0.9                      evaluate_0.14                
##  [85] xtable_1.8-4                  XML_3.99-0.4                 
##  [87] gridExtra_2.3                 compiler_4.0.0               
##  [89] biomaRt_2.45.2                tibble_3.0.3                 
##  [91] KernSmooth_2.23-17            crayon_1.3.4                 
##  [93] htmltools_0.5.0               mgcv_1.8-31                  
##  [95] later_1.1.0.1                 tidyr_1.1.0                  
##  [97] DBI_1.1.0                     dbplyr_1.4.4                 
##  [99] MASS_7.3-51.6                 rappdirs_0.3.1               
## [101] igraph_1.2.5                  pkgconfig_2.0.3              
## [103] pkgdown_1.5.1                 GenomicAlignments_1.25.3     
## [105] plotly_4.9.2.1                scuttle_0.99.11              
## [107] dqrng_0.2.1                   XVector_0.29.3               
## [109] stringr_1.4.0                 digest_0.6.25                
## [111] sctransform_0.2.1             RcppAnnoy_0.0.16             
## [113] spatstat.data_1.4-3           Biostrings_2.57.2            
## [115] rmarkdown_2.3                 leiden_0.3.3                 
## [117] uwot_0.1.8                    edgeR_3.31.4                 
## [119] DelayedMatrixStats_1.11.1     curl_4.3                     
## [121] shiny_1.5.0                   Rsamtools_2.5.3              
## [123] gtools_3.8.2                  lifecycle_0.2.0              
## [125] nlme_3.1-148                  jsonlite_1.7.0               
## [127] BiocNeighbors_1.7.0           desc_1.2.0                   
## [129] viridisLite_0.3.0             askpass_1.1                  
## [131] limma_3.45.9                  pillar_1.4.6                 
## [133] lattice_0.20-41               fastmap_1.0.1                
## [135] httr_1.4.1                    survival_3.2-3               
## [137] interactiveDisplayBase_1.27.5 glue_1.4.1                   
## [139] spatstat_1.64-1               png_0.1-7                    
## [141] BiocVersion_3.12.0            bit_1.1-15.2                 
## [143] stringi_1.4.6                 blob_1.2.1                   
## [145] BiocSingular_1.5.0            AnnotationHub_2.21.1         
## [147] memoise_1.1.0                 dplyr_1.0.0                  
## [149] irlba_2.3.3                   future.apply_1.6.0           
## [151] ape_5.4</code></pre>
</div>
<div id="references" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<div id="refs" class="references">
<div id="ref-pbmc">
<p>“1k Pbmcs from a Healthy Donor (V2 Chemistry).” n.d. <a href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.0/pbmc_1k_v2">https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.0/pbmc_1k_v2</a>.</p>
</div>
<div id="ref-Amezquita2020">
<p>Amezquita, Robert A., Aaron T. L. Lun, Etienne Becht, Vince J. Carey, Lindsay N. Carpp, Ludwig Geistlinger, Federico Marini, et al. 2020. “Orchestrating Single-Cell Analysis with Bioconductor.” <em>Nature Methods</em> 17 (2): 137–45. <a href="https://doi.org/10.1038/s41592-019-0654-x">https://doi.org/10.1038/s41592-019-0654-x</a>.</p>
</div>
<div id="ref-gencode">
<p>Frankish, A., GENCODE-consoritum, and P. Flicek. 2018. “GENCODE reference annotation for the human and mouse genomes.” <em>Nucleic Acids Research</em>. <a href="https://doi.org/10.1093/nar/gky955">https://doi.org/10.1093/nar/gky955</a>.</p>
</div>
<div id="ref-Lawrence2013">
<p>Lawrence, Michael, Wolfgang Huber, Hervé Pagès, Patrick Aboyoun, Marc Carlson, Robert Gentleman, Martin T. Morgan, and Vincent J. Carey. 2013. “Software for Computing and Annotating Genomic Ranges.” <em>PLOS Computational Biology</em> 9 (8): e1003118+. <a href="https://doi.org/10.1371/journal.pcbi.1003118">https://doi.org/10.1371/journal.pcbi.1003118</a>.</p>
</div>
<div id="ref-tximeta">
<p>Love, Michael I., Charlotte Soneson, Peter F. Hickey, Lisa K. Johnson, N. Tessa Pierce, Lori Shepherd, Martin Morgan, and Rob Patro. 2020. “Tximeta: Reference sequence checksums for provenance identification in RNA-seq.” <em>PLOS Computational Biology</em>. <a href="https://doi.org/10.1371/journal.pcbi.1007664">https://doi.org/10.1371/journal.pcbi.1007664</a>.</p>
</div>
<div id="ref-Lun2016">
<p>Lun, Aaron T. L., Karsten Bach, and John C. Marioni. 2016. “Pooling across cells to normalize single-cell RNA sequencing data with many zero counts.” <em>Genome Biology</em> 17 (1): 75. <a href="https://doi.org/10.1186/s13059-016-0947-7">https://doi.org/10.1186/s13059-016-0947-7</a>.</p>
</div>
<div id="ref-scran">
<p>Lun, A. T. L., D. J. McCarthy, and J. C. Marioni. 2016. “A Step-by-Step Workflow for Low-Level Analysis of Single-Cell Rna-Seq Data with Bioconductor.” <em>F1000Research</em> 5: 2122. <a href="https://doi.org/10.12688/f1000research.9501.2">https://doi.org/10.12688/f1000research.9501.2</a>.</p>
</div>
<div id="ref-salmon">
<p>Patro, R., G. Duggal, M. I. Love, R. A. Irizarry, and C. Kingsford. 2017. “Salmon Provides Fast and Bias-Aware Quantification of Transcript Expression.” <em>Nature Methods</em> 14: 417–19. <a href="https://doi.org/10.1038/nmeth.4197">https://doi.org/10.1038/nmeth.4197</a>.</p>
</div>
<div id="ref-Pijuan-Sala2019">
<p>Pijuan-Sala, Blanca, Jonathan A. Griffiths, Carolina Guibentif, Tom W. Hiscock, Wajid Jawaid, Fernando J. Calero-Nieto, Carla Mulas, et al. 2019. “A Single-Cell Molecular Map of Mouse Gastrulation and Early Organogenesis.” <em>Nature</em> 566 (7745): 490–95. <a href="https://doi.org/10.1038/s41586-019-0933-9">https://doi.org/10.1038/s41586-019-0933-9</a>.</p>
</div>
<div id="ref-alevin">
<p>Srivastava, Avi, Laraib Malik, Tom Sean Smith, Ian Sudbery, and Rob Patro. 2019. “Alevin efficiently estimates accurate gene abundances from dscRNA-seq data.” <em>Genome Biology</em> 20 (65). <a href="https://doi.org/10.1186/s13059-019-1670-y">https://doi.org/10.1186/s13059-019-1670-y</a>.</p>
</div>
<div id="ref-seurat">
<p>Stuart, Tim, Andrew Butler, Paul Hoffman, Christoph Hafemeister, Efthymia Papalexi, William M Mauck III, Yuhan Hao, Marlon Stoeckius, Peter Smibert, and Rahul Satija. 2019. “Comprehensive Integration of Single-Cell Data.” <em>Cell</em> 177: 1888–1902. <a href="https://doi.org/10.1016/j.cell.2019.05.031">https://doi.org/10.1016/j.cell.2019.05.031</a>.</p>
</div>
<div id="ref-compression">
<p>Van Buren, S, H Sarkar, A Srivastava, NU Rashid, R Patro, and MI Love. 2020. “Compression of quantification uncertainty for scRNA-seq counts.” <em>bioRxiv</em>. <a href="https://doi.org/10.1101/2020.07.06.189639">https://doi.org/10.1101/2020.07.06.189639</a>.</p>
</div>
<div id="ref-swish">
<p>Zhu, A., A. Srivastava, J. G. Ibrahim, R. Patro, and M. I. Love. 2019. “Nonparametric expression analysis using inferential replicate counts.” <em>Nucleic Acids Research</em>. <a href="https://doi.org/10.1093/nar/gkz622">https://doi.org/10.1093/nar/gkz622</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Michael Love, Avi Srivastava.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
